haskell/test/SMT/Common/Gen.hs

```diff
@@ -0,0 +1,424 @@
+    , cDecisionLambda =
+        toField #diExtra #
+        nil #
+        swap #
+        dip (push Nothing) #
+        constructStack @(DecisionLambdaOutput BigMap)
+    , cCustomEntrypoints = DynamicRec' $ mempty
+    , cFixedProposalFee = fixedProposalFee
+    , cPeriod = votingPeriod
+    , cProposalFlushLevel = (unPeriod votingPeriod) * 2
+    , cProposalExpiredLevel = (unPeriod votingPeriod) * 3
+    , cMaxQuorumChange = QuorumFraction $ percentageToFractionNumerator maxChangePercent
+    , cQuorumChange = QuorumFraction $ percentageToFractionNumerator changePercent
+    , cGovernanceTotalSupply = governanceTotalSupply
+    , cMaxQuorumThreshold = percentageToFractionNumerator 99 -- 99%
+    , cMinQuorumThreshold = percentageToFractionNumerator 1 -- 1%
+
+    , cMaxVoters = 1000
+    , cMaxProposals = 500
+    }
+
+genDelegates :: GeneratorT (Map Delegate ())
+genDelegates = do
+  userPool <- get <&> gsAddresses <<&>> fst
+  delegateList <- (Gen.shuffle $ (uncurry Delegate) <$> zip userPool userPool) >>= Gen.subsequence
+  pure $ Map.fromList $ zip delegateList $ repeat ()
+
+genDelegateParams :: GeneratorT [DelegateParam]
+genDelegateParams = do
+  userPool <- get <&> gsAddresses <<&>> fst
+  isEnable <- Gen.bool
+  userPool
+    <&> (\addr -> DelegateParam isEnable addr)
+     &  Gen.shuffle
+    >>= Gen.subsequence
+
+genAddressFreezeHistory :: GeneratorT AddressFreezeHistory
+genAddressFreezeHistory = do
+  lvl <- get <&> gsLevel
+  currentUnstaked <- Gen.integral (Range.constant 0 30)
+  pastUnstaked <- Gen.integral (Range.constant 0 40)
+  staked <- Gen.integral (Range.constant 0 50)
+  pure $ AddressFreezeHistory
+        { fhCurrentUnstaked = currentUnstaked
+        , fhPastUnstaked = pastUnstaked
+        , fhCurrentStageNum = lvl
+        , fhStaked = staked
+        }
+
+genFreezeHistory :: GeneratorT (Map Address AddressFreezeHistory)
+genFreezeHistory = do
+  addrs <- get <&> gsAddresses <<&>> fst
+  addrFreezeHistoryList <- vectorOf poolSize genAddressFreezeHistory
+  let freezeHistory = zip addrs addrFreezeHistoryList
+  pure $ Map.fromList freezeHistory
+
+genQuorumThresholdAtCycle :: Natural -> GeneratorT QuorumThresholdAtCycle
+genQuorumThresholdAtCycle staked = do
+  quorumThreshold <- Gen.integral (Range.constant 1 3)
+  lvl <- get <&> gsLevel
+  pure $ QuorumThresholdAtCycle lvl quorumThreshold staked
+
+genModelState :: SmtOption -> GeneratorT (Address -> Address -> ModelState)
+genModelState SmtOption{..} = do
+
+  mkStore <- genStorage
+  config <- genConfig
+  let mkFs = \guardian gov -> FullStorage' (mkStore guardian gov) config
+
+  selfAddrPlaceholder <- genAddress
+  pure $ \guardian gov ->
+    let fs = mkFs guardian gov
+    in ModelState
+        { msFullStorage = fs
+        , msMutez = toMutez 0
+
+        , msLevel = fs & fsStorage & sStartLevel
+        , msChainId = dummyChainId
+        , msSelfAddress = selfAddrPlaceholder -- This will be replace when dao is originated
```

Comment:
Can we use the same approach as `ModelInputArg`, instead of using a placeholder address?

Reply:
If we add a field to `ModelInputArg` for selfAddress, we then need to originate baseDAO first to create `ModelInputArg`. But to originate baseDAO, we need to get `modelState` which come from `mkModelInput` which needs `ModelInputArg`.

That's why we need this placeholder. It is probably possible to make `mkModelInput` return a `mkModelState` instead `modelState`, but that just complicates thing even more than it currently is.